### ISSUE TITLE
[Feature Request] Classifier-Free Guidance sampling

### ISSUE THREAD

**Issue by wronkiew** [mlc-ai/mlc-llm#499]
## 🚀 Feature
Add an option for models to use Classifier-Free Guidance (CFG) during inference. CFG uses a negative prompt to push inference to follow the system prompt more closely.

Support for this has also been requested at huggingface/transformers#24536 and ggerganov/llama.cpp#2083. The paper describing the technique is [here](https://arxiv.org/abs/2306.17806). Section 3.4 shows their evaluation using CFG to improve chatbot responses.

## Motivation
MLC-LLM is memory-constrained on mobile devices. The response quality of LLMs using CFG averaged similarly to standard models twice the size. For example, LLaMA-7B with CFG outperformed LLaMA-13B on the Lambada text-generation benchmark. This is at the expense of inference time.

**Comment by Vermeille** [comment]
Author here. Let me know if help is needed.

**Comment by yzh119** [comment]
Thank @wronkiew  for raising this up!
The example code by @Vermeille  is very intuitive, and the idea of contrasting logits w/ and w/o prompts looks interesting.

I suppose it needs some effort on the MLC-LLM side because we need to support multiple KV caches, I'm glad to help prototype it.
